### PR TITLE
Fixes #1429: Performance enhancement of backend connect with database by NIO in sharding-proxy

### LINKED ISSUE #1429: Performance enhancement of backend connect with database by NIO in sharding-proxy
I have researched currently program in sharding-proxy where implements backend connection with database by NIO, and I have some advisement for performance enhancement.
To show my idea clearly, I draw up design plan.
### Legend
Image of Legend

### Current case
Image of Current Case
Frontend and backend use different worker eventloop group(this means use different IO thread pool):

1. Application's request reaches worker eventloop group of frontend;
2. Then worker eventloop group writes data to user eventloop group, and the later processes protocol package; (Switch thread)
3. Then user eventloop group writes data to worker eventloop group of backend, and the later executes SQL(this means send protocol package to DB); (Switch thread)
4. Then query result from DB returned to work eventloop group of backend;
5. Then work eventloop group of backend writes data to user eventloop group; (Switch thread)
6. Then user eventloop group processes query result by blocking; 
7. Then user eventloop group writes the processed result to work eventloop group; (Switch thread)
8. Then work eventloop group writes back data to application.

Total 4 times switch thread in a query's life time.

### Case 1
Image of Current Case
Frontend and backend use same worker eventloop group(this means use same IO thread pool):

1. Application's request reaches worker eventloop group;
2. Then worker eventloop group writes data to user eventloop group, and the later processes protocol package; (Switch thread)
3. Then user eventloop group writes data to worker eventloop group, and the later executes SQL(this means send protocol package to DB); (Switch thread)
4. Then query result from DB returned to work eventloop group;
5. Then work eventloop group of backend writes data to user eventloop group; (Switch thread)
6. Then user eventloop group processes query result by blocking; 
7. Then user eventloop group writes the processed result to work eventloop group; (Switch thread)
8. Then work eventloop group writes back data to application.

Advantages: Reduce a thread pool.
Maybe only a little performance enhancement, but still 4 times switch thread in a query's life time.

### Case 2
Image of Current Case
Frontend and backend use same worker eventloop group(this means use same IO thread pool), and every eventloop in the group establish connection with DB by channel or channel pool:

1. Application's request reaches worker eventloop group;
2. Worker eventloop group processes protocol package and sends package to DB;(Considering processing protocol package in memory is very fast and no-blocking, so do this operation in IO thread to avoid switch thread)
3. Then query result from DB returned to work eventloop group;
4. If the query results need to be merged, switch thread and merge them in user eventloop group by blocking. If not, process them in IO thread pool to avoid switch thread;
5. If process the query results in user eventloop group, then user eventloop group writes data to work eventloop group(Switch thread);
6. Then work eventloop group writes back data to application.

Advantages:

1. Reduce a thread pool.
2. Reduce times of switch thread from 4 to 1 or 2.

Disadvantages: 

1. Every IO thread establishes connection with DB, if the count of IO thread in worker eventloop group is 4, then the count of connection is 4 times as the count of current case.

Could anyone give me some suggestions? Thank you very much!

### PR DESCRIPTION
**Pull request by solid210**
Fixes #1429.

Changes proposed in this pull request:
- Replace CountDownLatch with List and Count (means trade space for time), and remove SynchronizedFuture.
- Initialize channel pools for every eventLoop of frontend handler and data source name.
- If the query results need to be merged, use other thread pool to process for no-blocking.

### CODE REVIEW

File: sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/SchemaBroadcastBackendHandler.java [hunk 1]
```diff
@@ -39,6 +39,8 @@
 @RequiredArgsConstructor
 public final class SchemaBroadcastBackendHandler extends AbstractBackendHandler {
     
+    private static final GlobalRegistry GLOBAL_REGISTRY = GlobalRegistry.getInstance();
```
Comment (tuohai666) [inline sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/SchemaBroadcastBackendHandler.java]: is this necessary?

File: sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/ShowDatabasesBackendHandler.java [hunk 2]
```diff
@@ -41,6 +43,10 @@
  */
 public final class ShowDatabasesBackendHandler extends AbstractBackendHandler {
     
+    private static final GlobalRegistry GLOBAL_REGISTRY = GlobalRegistry.getInstance();
```
Comment (tuohai666) [inline sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/ShowDatabasesBackendHandler.java]: same as above.

File: sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/ShowDatabasesBackendHandler.java [hunk 3]
```diff
@@ -41,6 +43,10 @@
...
+    
+    private static final RuntimeContext RUNTIME_CONTEXT = RuntimeContext.getInstance();
```
Comment (tuohai666) [inline sharding-proxy/src/main/java/io/shardingsphere/shardingproxy/backend/ShowDatabasesBackendHandler.java]: same as above.

**Comment by terrymanu** [comment]
Closed. reason just ref #1893

**Pull request closed without merging**